tests/test_sharded_matmul.py

```diff
@@ -0,0 +1,251 @@
+        deepcopy(X),
+        deepcopy(Y),
+        replication_factor=num_ipus,
+        sharded_op=op,
+    )
+    lr = 1.0
+    if simulator.out_sharding == Sharding.Replicated:
+        lr /= num_ipus
+        optimizer = torch.optim.SGD(simulator.parameters(), lr=lr)
+    else:
+        params = [
+            {"params": simulator.X, "lr": lr},
+            {"params": simulator.Y, "lr": lr},
+        ]
+        if simulator.X_sharding == Sharding.Replicated:
+            params[0]["lr"] *= num_ipus  # type: ignore
+        if (
+            simulator.X_sharding == Sharding.Row
+            and simulator.Y_sharding == Sharding.Column
+        ):
+            params[0]["lr"] /= num_ipus  # type: ignore
+        optimizer = torch.optim.SGD(params, lr=lr)
+    optimizer.zero_grad()
+    out, loss = simulator()
+    loss.mean().backward()
+    grad_X = _sharding_transform_map[simulator.X_sharding]["out_sim"](simulator.X.grad)
+    grad_Y = _sharding_transform_map[simulator.Y_sharding]["out_sim"](simulator.Y.grad)
+    if simulator.X_sharding == Sharding.Replicated:
+        grad_X = einops.repeat(grad_X, "n d -> (r n) d", r=num_ipus)
+    if simulator.out_sharding != Sharding.Replicated:
+        grad_X *= num_ipus
+        grad_Y *= num_ipus
+    optimizer.step()
+    return out, grad_X, grad_Y, simulator.X.data, simulator.Y.data
+
+
+def run_sharded_matmul(
+    X: torch.Tensor, Y: torch.Tensor, op: Callable[[Any], torch.Tensor], num_ipus: int
+) -> Tuple[Any, Any, Any, Any, Any]:
+    options = poptorch.Options()
+    options.replicationFactor(num_ipus)
+    options.outputMode(poptorch.OutputMode.All)
+    options.useIpuModel(True)
+    options.anchorTensor("grad_X", "Gradient___X")
+    options.anchorTensor("grad_Y", "Gradient___Y")
+    options._Popart.setPatterns({"OpToIdentity": True})
+    tester = _ShardedMatmulTester(
+        deepcopy(X), deepcopy(Y), replication_factor=num_ipus, sharded_op=op
+    )
+    optimizer = poptorch.optim.SGD(tester.parameters(), lr=1.0)
+    tester = poptorch.trainingModel(tester, options, optimizer)
+    replicated_params = ["X"] if tester.X_sharding == Sharding.Replicated else []
+    utils._apply_replica_grouping(
+        tester, CommGroupType.Orthogonal, 1, excluded_parameters=replicated_params
+    )
+    out, _ = tester()
+    out = out.detach().cpu()
+    grad_X = tester.getAnchoredTensor("grad_X")  # type: ignore
+    grad_X = _sharding_transform_map[tester.X_sharding]["out_test"](grad_X, r=num_ipus)
+    grad_Y = tester.getAnchoredTensor("grad_Y")  # type: ignore
+    grad_Y = _sharding_transform_map[tester.Y_sharding]["out_test"](grad_Y, r=num_ipus)
+    return out, grad_X, grad_Y, tester.X.data, tester.Y.data
+
+
+@pytest.mark.parametrize("op", list(_op_mapping.keys()))
+def test_sharded_matmul(op: Callable[[Any], torch.Tensor]) -> None:
+    X, Y = generate_inputs()
+    num_ipus = 2
+    actual = run_sharded_matmul(X, Y, op, num_ipus)
+    expected = simulate_sharded_matmul(X, Y, op, num_ipus)
```

Comment:
I was going to suggest a plain matmul as reference, but now actually I wonder if it's useful to have both this & keeping the simulator. This would be pretty encouraging due to its simplicity.

Reply:
I did originally have this for forward pass only to check the simulated matmul produced the correct result (I'm not sure it would make sense with sharded/partial grads).